client/trino-client/src/main/java/io/trino/client/auth/external/Tokens.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.client.auth.external;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.airlift.json.JsonCodec;
+import io.airlift.units.Duration;
+import io.trino.client.ClientException;
+import io.trino.client.JsonResponse;
+import net.jodah.failsafe.Failsafe;
+import net.jodah.failsafe.FailsafeException;
+import net.jodah.failsafe.RetryPolicy;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Optional;
+
+import static io.airlift.json.JsonCodec.jsonCodec;
+import static io.trino.client.JsonResponse.execute;
+import static java.lang.String.format;
+import static java.net.HttpURLConnection.HTTP_OK;
+import static java.net.HttpURLConnection.HTTP_UNAVAILABLE;
+import static java.util.Objects.requireNonNull;
+
+public class Tokens
+{
+    private static final JsonCodec<TokenPollRepresentation> TOKEN_POLL_CODEC = jsonCodec(TokenPollRepresentation.class);
+
+    private final OkHttpClient client;
+
+    public Tokens(OkHttpClient client)
+    {
+        this.client = requireNonNull(client, "client is null");
+    }
+
+    public TokenPoll pollForTokenUntil(URI tokenUri, Duration timeout)
+            throws TokenPollException, ClientException
+    {
+        try {
+            return Failsafe.with(new RetryPolicy<TokenPoll>()
+                    .withMaxAttempts(-1)
+                    .withMaxDuration(java.time.Duration.ofMillis(timeout.toMillis()))
+                    .handleResultIf(TokenPoll::isPending)
+                    .abortOn(Throwable.class))
+                    .get(context -> {
+                        Request request = Optional.<TokenPoll>ofNullable(context.getLastResult())
+                                .flatMap(TokenPoll::getNextTokenUri)
+                                .map(Tokens::createRequestFor)
+                                .orElseGet(() -> createRequestFor(tokenUri));
+                        return handleRequest(request);
+                    });
+        }
+        catch (FailsafeException e) {
+            throw new TokenPollException(e.getCause());
+        }
+    }
+
+    private static Request createRequestFor(URI tokenUri)
+    {
+        try {
+            return new Request.Builder()
```

Comment:
Use `HttpUrl.get(uri)`, since `URL` is ancient and deprecated

We should also include `User-Agent` like we do in `StatementClientV1`